features/101-design-webhooks/design.md

```diff
@@ -0,0 +1,188 @@
+# WIP: next webhooks design
+
+## Overview
+
+skygear sends webhooks for events that happen in your app, and are designed to help a developer to monitor events, sync database, and many other use cases.
+
+When one of events is triggered, skygear sends a HTTP POST with a payload to the webhook's configured URL.
+
+## SYNC and ASYNC hooks
+
+There are two types of webhook:
+
+- `SYNC` - gear's operation will be blocked until a response is received. Gear should respect `SYNC` hook's response, if it indicates failed, operation should abort and return a formatted error response to the client.
+- `ASYNC` - gear's operation won't be blocked and the response from the webhook is ignored.
+
+## Request Headers
+
+Following headers will be added to webhook POST request:
+
+- `X-Skygear-Webhook-Req-ID`: an ID of the request.
+- `X-Skygear-Webhook-Async`: if is a `async` hook, then the value is `TRUE` otherwise is `FALSE`. 
+- `X-Skygear-Webhhok-Signature`: a signature string of the webhook event, a develop should can verify the request by skygear's SDK or manually.
+
+## Request Payload
+
+Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+
+All event payloads follows following payload format:
+
+```json=
+{
+    "event": "event_name",
+    "data": {
+        "key_name_1": "event_data_key_value",
+        "key_name_2": "event_data_key_value",
+        ...
+    }
+}
+```
+
+## Responding to a webhook
+
+To acknowledge receipt of a webhook, webhook endpoint should return a 2xx HTTP status code. All response codes outside this range, including 3xx codes, will indicate to skygear the request is failed.
+
+For some events, it may allow to webhook to modify changes (e.x. auth gear's before family hooks) when response code is 2xx. The behavior is defined by each gear, so will not be covered in this spec.
+
+## Configure webhook
+
+A gear should provide a REST interface for configuring webhooks, and a webhook is configured by following arguments:
+
+| argument | description | require |
+| -------- | -------- | -------- | 
+| `events` | A list of gear events, such as: ["after_signup", "before_login"]. | ✓ |
+| `url` | The url of the webhook. | ✓ |
+| `async` | default is `true`. | |
+| `secret` | default is empty, if provided, it will be used as the key to generate `X-Skygear-Webhhok-Signature` digest. | |
+
+
+Each gear should provide following REST interfaces:
```

Comment:
Does it mean that we will store the hooks information in the gear db? My original thought is storing it in the tenant config (gateway db), since now the gear db contains data not the setting. Cannot think of any pros and cons, maybe good too :)